framework/doc/content/source/kernels/ConservativeAdvection.md

```diff
@@ -22,27 +22,97 @@ element solution of the weak formulation. The first term is the volumetric term
 is a surface term describing the advective flux out of the
 volume. `ConservativeAdvection` corresponds to the former volumetric term.
 
-The corresponding Jacobian is given by
+Without numerical stabilization the corresponding Jacobian is given by
 
 \begin{equation}
 \frac{\partial R_i(u_h)}{\partial u_j} = -(\nabla \psi_i, \vec{v} \phi_j).
 \end{equation}
 
+## Full upwinding
+
+Advective flow is notoriously prone to physically-incorrect overshoots
+and undershoots, so in many simulations some numerical stabilization
+is used to reduce or eliminate this spurious behaviour.
+Full-upwinding [cite:dalen1979,adhikary2011] is an example of
+numerical stabilization and this essentially adds numerical diffusion
+to completely eliminate overshoots and undershoots.  Full-upwinding is
+available in `ConservativeAdvection` by setting the `upwinding_type`
+appropriately.
+
+For each element in the mesh, full upwinding is implemented in the
+following way.  For each $i$, the quantity
+\begin{equation}
+\tilde{R}_{i} = -(\nabla\psi_i, \vec{v}) \ ,
+\end{equation}
+is evaluated.  If $\tilde{R}_{i}>0$ is positive then mass (or heat, or whatever $u$
+represents) is flowing *out* of node $i$, and this is called an
+"upwind" node.  If $\tilde{R}_{i}\geq 0$, the residual for node $i$
+is set to
+\begin{equation}
+R_{i} = u_{i}\tilde{R}_{i} \ \ \texttt{ for } \ \ \tilde{R}_{i}\geq 0.
+\end{equation}
+Here $u_{i}$ is the value of $u$ at the node $i$.  Define the total
+mass flowing from the upwind nodes:
+\begin{equation}
+M_{\mathrm{out}} = \sum_{i\ \mathrm{ with }\ \tilde{R}_{i}\geq 0}
+u_{i}\tilde{R}_{i} \ .
+\end{equation}
+Similarly, define
+\begin{equation}
+\tilde{M}_{\mathrm{in}} = - \sum_{i \ \mathrm{ with }\  \tilde{R}_{i}< 0}
+\tilde{R}_{i} \ .
+\end{equation}
+Mass is conserved if the residual for the downwind nodes is defined to
+be
+\begin{equation}
+R_{i} = \tilde{R}_{i} M_{\mathrm{out}} / M_{\mathrm{in}}  \ \ \texttt{ for } \ \ \tilde{R}_{i}< 0.
+\end{equation}
+
+## Dirichlet boundary conditions for pure advection
+
+In the purely advective case, Dirichlet boundary conditions may be
+used to inject mass (or heat, or whatever $u$ corresponds to
+physically) into the domain, or extract it from the domain.
+Specifically:
+
+- if $\vec{v}\cdot\vec{n}<0$ (with $\vec{n}$ being the outward normal) then the velocity field is "blowing" mass into the domain.  Fixing $u=u_{B}$ on this boundary means the mass flux (units kg.s$^{-1}$.m$^{-2}$) entering the domain is $-\vec{v}\cdot\vec{n}u_{B}$.
+- if $\vec{v}\cdot\vec{n}>0$ (with $\vec{n}$ being the outward normal) then the velocity field is "blowing" mass out of the domain.  If no Dirichlet boundary condition is used then mass is prevented from leaving the domain via this boundary: mass arriving here will "build up", causing $u$ to increase.  Alternatively, if $u=0$ is fixed at the boundary all mass arriving here will exit the domain immediately.
```

Comment:
Should we add an integrated BC on downwind: $\langle\psi_i, \vec{v} u
 \cdot \vec{n}\rangle$ (from the integration by part above)?

Reply:
Yes, you are probably correct.  i'll investigate next week.

Reply:
Actually, on second thoughts, i wonder whether this is necessary.  Fixing u at the boundary naturally does all the integrated stuff, doesn't it?

Reply:
So now we have three options: Dirichlet (specifying u), Integrated (using $\langle\psi_i, \vec{v} u
\cdot \vec{n}\rangle$), or none. I do not know how they are related.

Reply:
I typically use `IntegratedBC` to apply BC both on upwinding and downwinding. Just split $\langle\psi_i, \vec{v} u \cdot \vec{n}\rangle$ into two parts, for the upwinding, substitute u_B into u, for the dowdwinding, keep u. u=u_B will be a steady-state solution. And there is no discontinuity in the solution on the downwinding boundary like applying Dirichlet boundary there.

Reply:
Here's my understanding of the BCs (also related to your comment below).  I'll rewrite the doco to reflect this, once you confirm that it agrees with your understanding.  Thanks @YaqiWang .   Basically, i think i was stupid in introducing DirichletBCs into the discussion.

To form the correct equations, we always need to include the <v.n u> in the equations (n is outward normal to the boundary).  In MOOSE we do this by always including a BC, which is called OutflowBC along every boundary.   Physically what this does is subtract away <v.n u> "fluid mass" from the boundary.

So for v.n>0 adding the OutflowBC allows "fluid" to flow freely through the boundary.  The velocity is "blowing fluid" into this boundary, and the OutflowBC is removing it at the correct rate, because the flux through any area is <v.n u>.

For v.n<0 adding the OutflowBC actually ends up fixing u at the boundary to its initial condition.  This is because the ConservativeAdvection Kernel is making fluid flow from the boundary to the interior of the model, but at the same time the OutflowBC is removing <v.n u> : note that v.n<0, so the OutflowBC is actually adding fluid at the same rate the ConservativeAdvection Kernel is adding it.  The physical interpretation is that something external to the model is adding fluid at exactly the rate specified by the initial conditions at that boundary.

Usually this isn't what we want for v.n<0.  Usually we want to specify a particular value for the injected flux, and we do that using InflowBC, which is just adding <v.n u_B>, where u_B (kg/m^2/s) is the injection rate.  (Using a DirichletBC to do this is just **wrong**, sorry.)

To make impermeable boundaries, either for v.n>0 or v.n<0 simply use no BC at that boundary.  Then for v.n>0 there is no BC to remove fluid from that boundary so the fluid "piles up" there.  For v.n<0 the omission of a BC can be thought of setting u_B=0 in an InflowBC.  (Again, using a DirichletBC is **wrong**, sorry).

Reply:
Thanks Andy. Did you updated the doc to reflect your comment here?